File: rb/features.rb
# CustomsWindow SDK feature factory

require_relative 'feature/base_feature'
require_relative 'feature/debug_feature'
require_relative 'feature/idempotency_feature'
require_relative 'feature/metrics_feature'
require_relative 'feature/paging_feature'
require_relative 'feature/ratelimit_feature'
require_relative 'feature/retry_feature'
require_relative 'feature/test_feature'
require_relative 'feature/timeout_feature'


module CustomsWindowFeatures
  def self.make_feature(name)
    case name
    when "base"
      CustomsWindowBaseFeature.new
    when "debug"
      CustomsWindowDebugFeature.new
    when "idempotency"
      CustomsWindowIdempotencyFeature.new
    when "metrics"
      CustomsWindowMetricsFeature.new
    when "paging"
      CustomsWindowPagingFeature.new
    when "ratelimit"
      CustomsWindowRatelimitFeature.new
    when "retry"
      CustomsWindowRetryFeature.new
    when "test"
      CustomsWindowTestFeature.new
    when "timeout"
      CustomsWindowTimeoutFeature.new
    else
      CustomsWindowBaseFeature.new
    end
  end
end

File: py/customswindow_sdk/features.py
# CustomsWindow SDK feature factory

from customswindow_sdk.feature.base_feature import CustomsWindowBaseFeature
from customswindow_sdk.feature.debug_feature import CustomsWindowDebugFeature
from customswindow_sdk.feature.idempotency_feature import CustomsWindowIdempotencyFeature
from customswindow_sdk.feature.metrics_feature import CustomsWindowMetricsFeature
from customswindow_sdk.feature.paging_feature import CustomsWindowPagingFeature
from customswindow_sdk.feature.ratelimit_feature import CustomsWindowRatelimitFeature
from customswindow_sdk.feature.retry_feature import CustomsWindowRetryFeature
from customswindow_sdk.feature.test_feature import CustomsWindowTestFeature
from customswindow_sdk.feature.timeout_feature import CustomsWindowTimeoutFeature


_FEATURES = {
    "base": lambda: CustomsWindowBaseFeature(),
    "debug": lambda: CustomsWindowDebugFeature(),
    "idempotency": lambda: CustomsWindowIdempotencyFeature(),
    "metrics": lambda: CustomsWindowMetricsFeature(),
    "paging": lambda: CustomsWindowPagingFeature(),
    "ratelimit": lambda: CustomsWindowRatelimitFeature(),
    "retry": lambda: CustomsWindowRetryFeature(),
    "test": lambda: CustomsWindowTestFeature(),
    "timeout": lambda: CustomsWindowTimeoutFeature(),
}


def _make_feature(name):
    factory = _FEATURES.get(name)
    if factory is not None:
        return factory()
    return _FEATURES["base"]()


# True when this SDK was generated with the named feature class - the
# constructor's tolerance for extend-carried features reads this (an
# active name with no generated class must not become a BaseFeature
# stray when an extend instance carries it).
def _has_feature(name):
    return name in _FEATURES
